Pad server actuator state

bskdany opened the pull request:
- Added actuator state tracking on pad server
- Fixed arming progression based on the spec

In actuator.c line 116 I'm unsure if I should return ACT_OK or something ACT_INV
```
 if (new_state == current_state) {
        return ACT_OK;
    }
```

linguini1 commented on pad_server/src/actuator.c:
```diff
@@ -52,4 +55,73 @@ int actuator_off(actuator_t *act) { return act->off(act); }
+        return -1;
+    }
+
+    bool is_solanoid_valve = id >= ID_XV1 && id <= ID_XV12;
```
Nit but it's spelt "solenoid".

linguini1 commented on pad_server/src/state.c:
```diff
@@ -52,3 +53,15 @@ int padstate_actuate(padstate_t *state, uint8_t id, bool new_state) {
     return pthread_rwlock_unlock(&state->rw_lock);
 }
+
+int padstate_get_actuator(padstate_t *state, uint8_t act_id, bool *act_val) {
```
Nit but I'd call this `padstate_get_actstate` or something since you're not getting the actual actuator, just it's
state.

linguini1 reviewed (changes requested):
All the logic looks excellent to me, just make sure your `actuator_set` function actually calls the actuator's on/off functions to perform the setting.

bskdany commented on pad_server/src/actuator.h:
```diff
@@ -42,6 +43,7 @@ typedef struct actuator {
 int actuator_on(actuator_t *act);
 int actuator_off(actuator_t *act);
 void actuator_init(actuator_t *act, uint8_t id, actuate_f on, actuate_f off, void *priv);
+int actuator_set(padstate_t *state, uint8_t id, uint8_t new_state);
```
I'm trying to see the vision for the architecture here:

The ` actuator_set` function gets the actuator state, checks if the new state is different than the current state and calls either the on and off function.
Internally the on and off functions call `padstate_actuate`, in the future they will have additional logic for each actuator, that's why an array of the `actuator_t` type is needed.

I don't get this part
> and whether or not the request is OK/DENIED.

Does this mean the validity check for the actuation request needs to be checked in the on/off functions? I'm confused.

linguini1 replied:
For the architecture, the goal is that each actuator will provide only two functions: on and off. Those are what are used to control the actuator.

Since the command communication spec allows setting of the actuator state itself, I think it's fair to have an `actuator_set` command like you implemented. However, this shouldn't be coupled with the logic for arming, it should just call either on or off functions based on the requested state.

```c
// Something like this?
int actuator_set(act_t *act, bool requested_state) {
    if (requested_state) {
        return act->on(act);
    }
    return act->off(act);
}
```

Instead, the arming checks should take place in something like the `padstate_actuate` function. This function checks to make sure the actuation is permitted (DENIED/OK), and if it is permitted then it actually calls `actuator_set` on the actuator. All the actuator* commands should blindly execute and not perform arming logic error checking, that will be a calling code responsibility. Basically move the switch statement into another function and have it verify that the actuation would be allowed before performing it

I believe the current implementation has `actuator_set` performing the logic checks. Any functions that are prefixed with `actuator` should, in my opinion, be limited to only operating on the actuator and not requiring knowledge of the arming state. If you have a different idea for the architecture I would be happy to hear it though!

bskdany replied:
Ok I see. So the chain of command would be
controller -> padstate_actuate (does the checks, retrieves the actuator) -> actuator_set -> on (operations with pthread)

If that's the case it's kinda awkward because the other functions starting with padstate lock the state directly, but I can work around that and change the naming here and there.

linguini1 replied:
The `padstate_actuate` will have to lock the state to check the arming level, so not really far removed! But yep, that's the chain of command. I don't think `on` does any pthread operations though.

bskdany commented:
Since last review:
- controller_client now calls padstate_actuate directly, which checks the actuator permissions and then calls actuator_set.
- actuator_set changes the state, then it retrieves the actuator_t and calls the on function, which was initialized to be empty for now, as no physical actuators are attached yet.

linguini1 commented on pad_server/src/state.c:
```diff
@@ -40,15 +43,85 @@ int padstate_change_level(padstate_t *state, arm_lvl_e new_arm) {
     if (err) return err;
 
-    state->actuators[id] = new_state;
+    *act_state = state->actuators[act_id];
```
This is exactly how we should be reading state. With an actuator implementation, you can replace the `bool
actuators[NUM_ACTUATORS]` with a `actuator_t actuators[NUM_ACTUATORS]`. Then you can do your
`actuator_controllers_init()` logic inside the `padstate_init` function.

bskdany replied:
Ok got it. I would have to modify actuator_t to have a state field correct?

linguini1 replied:
Yes I forgot it currently doesn't have one!

bskdany replied:
Ok, that explains a lot of things

linguini1 commented on pad_server/src/actuator.c:
```diff
@@ -38,18 +49,42 @@ void actuator_init(actuator_t *act, uint8_t id, actuate_f on, actuate_f off, voi
+ * @return ACT_OK for success, ACT_DNE for invalid id, ACT_INV for invalid req_state, -1 for errors eith errno being
+ * set
+ */
+int actuator_set(padstate_t *state, uint8_t id, bool new_state) {
```
This function doesn't need the state, just the pointer to an actuator. It should operate only on the actuator.

If you want to operate on the state you should create a function in the padstate to call `actuator_set` after locking the wrlock.

bskdany commented:
Here we go again:
- added a state field to actuator_t.
- added a `pad_actuate` function in `state.c` that handles the permission logic.
- removed `actuator_set` in favor of `actuator_on` and `actuator_off` as it would have been left unused.

linguini1 reviewed (commented):
Few nitpicks about struct opacity which we discussed, but you also mentioned a great point about huge critical sections with my design plan. I am merging this since the progression logic is great, and you'll probably have a PR open shortly to implement the better locking we discussed soon anyways. Thanks as always, you're really making `hysim` usable!
